File: java/com/craftinginterpreters/vox/GenerateAst.java
//>= Syntax Trees
package com.craftinginterpreters.vox;

import java.io.IOException;
import java.io.PrintWriter;
import java.util.Arrays;
import java.util.List;

public class GenerateAst {
  public static void main(String[] args) throws IOException {
    defineAstType("Expr", Arrays.asList(
        // TODO: Don't have object until Classes chapter.
//>= Variables
        "Assign   : Expr object, Token name, Expr value",
//>= Syntax Trees
        "Binary   : Expr left, Token operator, Expr right",
//>= Functions
        "Call     : Expr callee, Token paren, List<Expr> arguments",
//>= Syntax Trees
        "Grouping : Expr expression",
        "Literal  : Object value",
//>= Control Flow
        "Logical  : Expr left, Token operator, Expr right",
//>= Classes
        "Property : Expr object, Token name",
//>= Inheritance
        "Super    : Token keyword, Token method",
//>= Classes
        "This     : Token name",
//>= Syntax Trees
        "Unary    : Token operator, Expr right",
//>= Variables
        "Variable : Token name"
//>= Syntax Trees
    ));
//>= Variables

    defineAstType("Stmt", Arrays.asList(
        "Block       : List<Stmt> statements",
//>= Classes
        "Class       : Token name, Expr superclass, List<Stmt.Function> methods",
//>= Variables
        "Expression  : Expr expression",
//>= Uhh
        "For         : Token name, Expr iterator, Stmt body",
//>= Functions
        "Function    : Token name, List<Token> parameters, Stmt body",
//>= Control Flow
        "If          : Expr condition, Stmt thenBranch, Stmt elseBranch",
//>= Functions
        "Return      : Token keyword, Expr value",
//>= Variables
        "Var         : Token name, Expr initializer",
//>= Control Flow
        "While       : Expr condition, Stmt body"
//>= Variables
    ));
//>= Syntax Trees
  }

  private static void defineAstType(String baseName, List<String> types)
      throws IOException {
    PrintWriter writer = new PrintWriter(
        "com/craftinginterpreters/vox/" + baseName + ".java", "UTF-8");

    writer.println("package com.craftinginterpreters.vox;");
    writer.println("");
    writer.println("import java.util.List;");
    writer.println("");
    writer.println("abstract class " + baseName + " {");

    defineVisitor(writer, baseName, types);

    // The AST classes.
    for (String type : types) {
      String className = type.split(":")[0].trim();
      String fields = type.split(":")[1].trim();
      defineType(writer, baseName, className, fields);
    }

    // The base accept() method.
    writer.println("");
    writer.println("  abstract <R> R accept(Visitor<R> visitor);");

    writer.println("}");
    writer.close();
  }

  private static void defineVisitor(
      PrintWriter writer, String baseName, List<String> types) {
    writer.println("  interface Visitor<R> {");

    for (String type : types) {
      String typeName = type.split(":")[0].trim();
      writer.println("    R visit" + typeName + baseName + "(" +
          typeName + " " + baseName.toLowerCase() + ");");
    }

    writer.println("  }");
  }

  private static void defineType(PrintWriter writer, String baseName,
                                 String className, String fieldList) {
    writer.println("");
    writer.println("  static class " + className + " extends " +
        baseName + " {");

    // Constructor.
    writer.println("    " + className + "(" + fieldList + ") {");

    // Store parameters in fields.
    String[] fields = fieldList.split(", ");
    for (String field : fields) {
      String name = field.split(" ")[1];
      writer.println("      this." + name + " = " + name + ";");
    }

    writer.println("    }");

    // Visitor pattern.
    writer.println();
    writer.println(
        "    <R> R accept(Visitor<R> visitor) {");
    writer.println("      return visitor.visit" + className +
        baseName + "(this);");
    writer.println("    }");

    // Fields.
    writer.println();
    for (String field : fields) {
      writer.println("    final " + field + ";");
    }

    writer.println("  }");
  }
}

File: python/pylox/pylox/generate_ast.py
import os
import sys


def main():
    args = sys.argv[1:]
    if len(args) != 1:
        print("Usage: generate_ast.py <output_dir>")
        sys.exit(64)
    output_dir = args[0]

    define_ast(output_dir, "Expr", [
        "Assign : Token name, Expr value",
        "Binary : Expr left, Token operator, Expr right",
        "Grouping : Expr expression",
        "Literal : object value",
        "Unary : Token operator, Expr right",
        "Variable : Token name",
    ])

    define_ast(output_dir, "Stmt", [
        "Block : List[Stmt] statements",
        "If : Expr condition, Stmt then_branch, Stmt else_branch",
        "Expression : Expr expression",
        "Print : Expr expression",
        "Var : Token name, Expr initializer",
    ])


def define_ast(output_dir, basename, types):
    path = os.path.join(output_dir, f"{basename.lower()}.py")
    with open(path, "w") as f:
        f.write(f"""
import abc
from dataclasses import dataclass
{"from typing import List" if basename == "Stmt" else ""}
{"from pylox.expr import Expr" if basename == "Stmt" else ""}
from pylox.tokens import Token


class {basename}:
    def accept(self, visitor: 'Visitor'):
        raise NotImplementedError
""")

        for type in types:
            classname, fields = [x.strip() for x in type.split(":")]
            define_type(f, basename, classname, fields)

        define_visitor(f, basename, types)


def define_type(f, basename, classname, fields):
    f.write(f"""

@dataclass
class {classname}({basename}):
""")

    for field in fields.split(", "):
        type, name = field.split(" ")
        f.write(f"    {name}: {type}\n")

    f.write(f"""
    def accept(self, visitor: 'Visitor'):
        return visitor.visit_{classname.lower()}(self)
""")


def define_visitor(f, basename, types):
    f.write("""

class Visitor(metaclass=abc.ABCMeta):
""")
    for type in types:
        typename = type.split(":")[0].strip()
        f.write(f"""
    @abc.abstractmethod
    def visit_{typename.lower()}(self, {basename.lower()}: {typename}):
        raise NotImplementedError
""")


if __name__ == "__main__":
    main()
